Introduce "async" rpc handlers and use them in the raft transport.

bdarnell opened the pull request:
This allows us to preserve order between raft messages that were sent
over the same connection (out of order raft messages were a performance
problem because they cause unnecessary retries).

Closes #1669.

bdarnell committed:
Introduce "async" rpc handlers and use them in the raft transport.

This allows us to preserve order between raft messages that were sent
over the same connection (out of order raft messages were a performance
problem because they cause unnecessary retries).

Closes #1669.

petermattis commented on rpc/server.go:
```diff
@@ -82,12 +90,26 @@ func NewServer(addr net.Addr, context *Context) *Server {
-// Both the argument and return value of 'handler' should be a pointer to a protocol
-// message type.
+// Register a new method handler. `name` is a qualified name of the
+// form "Service.Name".  `handler` is a function that takes an
```
I thought two spaces after a period died with the typewriter.

bdarnell replied:
Old habits die hard. I wish we had a `commentfmt` so no one has to waste time thinking about this.

bdarnell replied:
Some of these aren't even my fault. Emacs users: add `(setq sentence-end-double-space nil)` or emacs may add extra spaces when you format comments with `M-q`. Now I just need to figure out how to write a save hook that runs `canonically-space-region` on every comment.

petermattis commented on rpc/server.go:
```diff
@@ -82,12 +90,26 @@ func NewServer(addr net.Addr, context *Context) *Server {
+}
+
+// RegisterAsync registers an asynchronous method handler. Instead of
+// returning an (interface{}, error) tuple, an asynchronous handler
```
`s/interface{}/proto.Message/g`

petermattis commented on multiraft/multiraft.go:
```diff
@@ -33,6 +33,8 @@ import (
 const (
 	noGroup = proto.RaftID(0)
+
+	reqBufferSize = 100
```
Is this large enough? You mention in the `RegisterAsync` comments that the handler should not block, but `multiraftServer.RaftMessage` will block if the request channel is full. The alternative to blocking there is to change `multiraftServer` to not use a channel but to have a list of requests and a condition variable.

bdarnell replied:
A list+condvar is essentially a channel with unlimited size, which is kind of scary. Also, we need to be able to `select` on this channel so if we used a list+condvar we'd need a separate goroutine to copy messages from the list onto an unbuffered channel as they become available.

Blocking here is not the end of the world; it just results in backpressure on the TCP connection, so it might be reasonable to just let it block when the channel fills up. The biggest risk of blocking in the rpc thread is deadlock, if we can't make progress due to something that indirectly depends on some future RPC getting through to this node. But I don't *think* that can happen with raft messages; we should always be able to clear any backlog at the raft level without any non-local resources. 

I think that as long as we only block the rpc thread in exceptional circumstances then the bounded channel is probably the best solution. If this does turn out to be a problem then I think I'd rather start dropping messages than move to buffering them in an unbounded list.

petermattis commented:
LGTM

tamird commented on server/raft_transport_test.go:
```diff
@@ -142,3 +143,72 @@ func TestSendAndReceive(t *testing.T) {
+
+// TestInOrderDelivery verifies that for a given pair of nodes, raft
+// messages are delivered in order. This test is unfortunately not
+// very sensitive; even when we were not guaranteeing in-order
```
would sprinkling random sleeps help with the sensitivity?

tamird replied:
(or non random ones)

bdarnell replied:
Yes. I initially tested this change with (non-committed) changes to the rpc server itself to introduce delays, but I didn't want to commit that kind of change to non-test code. But I've added a new commit that does something similar in the test-only channelServer.

tamird replied:
:+1: probably should update this comment

bdarnell committed:
Introduce random sleeps to the TestInOrderDelivery test.

With this change (and `-cpus 4`), the test reliably reproduces
reorderings without the previous commit (and reliably passes with it).

tamird commented on server/raft_transport_test.go:
```diff
@@ -119,13 +136,14 @@ func TestSendAndReceive(t *testing.T) {
-	// Read all the messages from the channels.  Note that the transport does not
-	// currently guarantee in-order delivery, so we just verify that the right number
-	// of messages end up in each channel.
+	// Read all the messages from the channels.  Note that the transport
```
two spaces? :trollface:

tamird commented on server/raft_transport.go:
```diff
@@ -87,13 +90,19 @@ func (t *rpcTransport) RaftMessage(args gogoproto.Message) (gogoproto.Message, e
-	// multiraft's internal struct and the external proto representation.
+	// Raft responses are empty so we don't actually need to convert
+	// between multiraft's internal struct and the external proto
+	// representation.  In fact, we don't even need to wait for the
```
:trollface:

tamird commented on rpc/server.go:
```diff
@@ -45,6 +45,14 @@ type serverResponse struct {
 	err   error
 }
 
+type syncAdapter func(proto.Message) (proto.Message, error)
```
should this be called asyncAdapter? this name surprised me

bdarnell replied:
Maybe; I don't feel strongly either way. I'm not happy with the use of "sync" and "async" here in general: is a handler that takes a callback asynchronous because it may not finish until some time after it has returned, or is it synchronous because it is called synchronously with respect to the rpc server's thread?

tamird commented:
LGTM. This came together pretty nicely.

tschottdorf commented:
LGTM

spencerkimball commented:
LGTM
